cluster/gce/gci/configure-helper.sh

```diff
@@ -2693,11 +2693,11 @@ function main() {
   fi
 
   override-kubectl
+  start-kubelet
```

Comment:
I think this may cause unnecessary complications if kubelet starts up super fast and immediately starts running containers before docker is configured correctly and restarted.

Is this "optimization" really needed?

Reply:
Yes, but docker will be in the terminating state, and kubelet wouldn't be able to create new containers through it.

Just to be clear, I don't think your change would cause issues today, but this *may* open a door for race conditions in the future (and that'd be hard to catch and safeguard against). Are there any other low-hanging fruits besides this one?

Reply:
This is quite cheap and saves around 10% of node start up. If it is possible I want to fix all low-hanging fruits.

Reply:
I'm not sure if I understand your concerns. There is no guaranties that docker is started before kubelet right now. We already have  race conditions. If this is a problem then shouldn't we fix that? If not then my change shouldn't break anything.

Reply:
`assemble-docker-flags` configures docker and restarts it.

What I'd like to prevent is that kubelet starts up and talks to the older docker instance (before restart, with the wrong configuration). Your PR increases the risk of that.

Reply:
I see the problem.
I tested another version of my changes where:
1. docker is stopped
2. kubelet started
3. docker started
but it saves only 1 second.
Do you see any other way to restart docker after kubelet is started?
Is it possible not to start docker with incorrec flags so we would start it instead of restarting?
Maybe kubelet could wait till docker is restarted?

Reply:
IIRC, kubelet will not start up at all unless docker is already up. It probably exited right away, and crashed loop until docker was up. That could partially explain why only 1s was saved.

The OS images come with docker pre-installed and running. The script makes sure that we configure the docker properly, as we want it. To avoid this, you'll have to bake the configurations into the OS image itself. As for kubelet, it doesn't know/care about how docker is configured (those are vendor-specific), so it can't wait for the right docker instance to be up.

Reply:
I saved in my test only 1 second (vs 2-3) because it takes 1-2 seconds to stop docker. In previous test docker was stopped at the same time as kubelet started. This is why I prefer solution for this PR.

Could you verify if kubelet is checking if docker is up? Is it a big problem if pod is started on docker before restart, won't it be started again when docker is restarted?

Reply:
> Is it a big problem if pod is started on docker before restart, won't it be started again when docker is restarted?

I don't think the 2 second optimization is a good reason to run and restart users containers. Doing so may also cause complications, e.g., RestartPolicy=Never pods may not be restarted properly. If docker has live-restore enabled, the containers may not be killed, which could lead to more complications.